test/test.awsrequestsigner.ts

```diff
@@ -0,0 +1,297 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as assert from 'assert';
+import {describe, it, afterEach, beforeEach} from 'mocha';
+import * as sinon from 'sinon';
+import {AwsRequestSigner} from '../src/auth/awsrequestsigner';
+import {GaxiosOptions} from 'gaxios';
+
+/** Defines the interface to facilitate testing of AWS request signing. */
+interface AwsRequestSignerTest {
+  // Test description.
+  description: string;
+  // AWS request signer instance.
+  instance: AwsRequestSigner;
+  // The raw input request.
+  originalRequest: GaxiosOptions;
+  // The expected signed output request.
+  getSignedRequest: () => GaxiosOptions;
+}
+
+describe('AwsRequestSigner', () => {
+  let clock: sinon.SinonFakeTimers;
+  // Load AWS credentials from a sample security_credentials response.
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const awsSecurityCredentials = require('../../test/fixtures/aws-security-credentials.json');
+  const referenceDate = new Date('2020-08-11T06:55:22.345Z');
+  const amzDate = '20200811T065522Z';
+  const dateStamp = '20200811';
+  const awsRegion = 'us-east-2';
+  const accessKeyId = awsSecurityCredentials.AccessKeyId;
+  const secretAccessKey = awsSecurityCredentials.SecretAccessKey;
+  const token = awsSecurityCredentials.Token;
+
+  beforeEach(() => {
+    clock = sinon.useFakeTimers(referenceDate);
+  });
+
+  afterEach(() => {
+    if (clock) {
+      clock.restore();
+    }
+  });
+
+  describe('getRequestOptions()', () => {
+    const awsError = new Error('Error retrieving AWS security credentials');
+    // Successful AWS credentials retrieval.
+    // In this case, temporary credentials are returned.
+    const getCredentials = async () => {
+      return {
+        accessKeyId,
+        secretAccessKey,
+        token,
+      };
+    };
+    // Successful AWS credentials retrieval.
+    // In this case, permanent credentials are returned (no session token).
+    const getCredentialsWithoutToken = async () => {
+      return {
+        accessKeyId,
+        secretAccessKey,
+      };
+    };
+    // Failing AWS credentials retrieval.
+    const getCredentialsUnsuccessful = async () => {
+      throw awsError;
+    };
+    // Sample request parameters.
+    const requestParams = {
+      KeySchema: [
+        {
+          KeyType: 'HASH',
+          AttributeName: 'Id',
+        },
+      ],
+      TableName: 'TestTable',
+      AttributeDefinitions: [
+        {
+          AttributeName: 'Id',
+          AttributeType: 'S',
+        },
+      ],
+      ProvisionedThroughput: {
+        WriteCapacityUnits: 5,
+        ReadCapacityUnits: 5,
+      },
+    };
+    // List of various requests and their expected signatures.
+    // Examples source:
+    // https://docs.aws.amazon.com/general/latest/gr/sigv4-signed-request-examples.html
+    const getRequestOptionsTests: AwsRequestSignerTest[] = [
```

Comment:
I like this DSL you've built out for exercising the the fairly complex logic in the AWS request signer.

Reply:
Thank you. I think this simplifies adding and removing new tests for this logic. It also makes it easier to replace the existing canonical request/signatures we are using. So it should not be expensive to maintain and build on in the future.